crates/driver/src/tests/setup/mod.rs

```diff
@@ -1,12 +1,988 @@
+        let expected = match fulfillment.quote.order.side {
+            order::Side::Buy => {
+                (fulfillment.quote.sell - fulfillment.quote.order.surplus_fee()).to_string()
+            }
+            order::Side::Sell => fulfillment.quote.buy.to_string(),
+        };
+        assert_eq!(amount, expected);
+        self
+    }
+
+    /// Check that the quote returns the expected amount of tokens. This is
+    /// based on the state of the blockchain and the test setup.
+    pub fn interactions(self) -> Self {
+        assert_eq!(self.fulfillments.len(), 1);
+        let fulfillment = &self.fulfillments[0];
+        let result: serde_json::Value = serde_json::from_str(&self.body).unwrap();
+        let interactions = result
+            .get("interactions")
+            .unwrap()
+            .as_array()
+            .unwrap()
+            .to_owned();
+        assert_eq!(interactions.len(), fulfillment.interactions.len());
+        for (interaction, expected) in interactions.iter().zip(&fulfillment.interactions) {
+            let target = interaction.get("target").unwrap().as_str().unwrap();
+            let value = interaction.get("value").unwrap().as_str().unwrap();
+            let calldata = interaction.get("callData").unwrap().as_str().unwrap();
+            assert_eq!(target, format!("0x{}", hex::encode(expected.address)));
+            assert_eq!(value, "0");
+            assert_eq!(calldata, format!("0x{}", hex::encode(&expected.calldata)));
+        }
+        self
+    }
+}
+
+/// The expected difference between a previous user balance for a certain token
+/// and the balance after the settlement has been broadcast.
+#[derive(Debug, Clone, Copy)]
+pub enum Balance {
+    /// The balance should be greater than before.
+    Greater,
+    /// The balance should be smaller than before by an exact amount.
+    SmallerBy(eth::U256),
+}
+
+/// A /settle response.
+pub struct Settle<'a> {
+    old_balances: HashMap<&'static str, eth::U256>,
+    status: StatusCode,
+    test: &'a Test,
+    solution_id: String,
+}
+
+pub struct SettleOk<'a> {
+    test: &'a Test,
+    old_balances: HashMap<&'static str, eth::U256>,
+}
+
+impl<'a> Settle<'a> {
+    /// Expect the /settle endpoint to have returned a 200 OK response.
+    pub async fn ok(self) -> SettleOk<'a> {
+        // Ensure that the response is OK.
+        assert_eq!(self.status, hyper::StatusCode::OK);
+
+        // Ensure that the solution ID is included in the settlement.
+        let tx = self
+            .test
+            .blockchain
+            .web3
+            .eth()
+            .transaction(TransactionId::Block(
+                BlockId::Number(ethcontract::BlockNumber::Latest),
+                0.into(),
+            ))
+            .await
+            .unwrap()
+            .unwrap();
+        let input = tx.input.0;
+        let len = input.len();
+        let tx_solution_id = u64::from_be_bytes((&input[len - 8..]).try_into().unwrap());
```

Comment:
Actually, nowadays this _is_ guaranteed. Since the driver submits the settlement, there has to be a way to the autopilot to recognize it. This happens via the settlement ID. Here's the [related domain type](https://github.com/cowprotocol/services/blob/dec4abfb0f540344c5337e125e7f210d6826d85d/crates/driver/src/domain/competition/solution/settlement.rs#L326-L332).

Reply:
Do you think I should improve the doc comment to mention that it's encoded _at the end_ of the calldata?

Reply:
I know that the driver already does this at the moment. My point was if we were to change that logic and introduce a bug which for example adds 1 byte less this code would not pick the correct bytes as the id to compare against. In that scenario it would make the last byte of the actual call data part of the id to compare against and **might not** detect that regression.
However, the chance that miraculously all these ids match by accident to please the CI pipeline is probably very low.